### DISCUSSION TITLE
Mixin Overwriting a Public Static Method

### DISCUSSION THREAD

**Discussion by crosby-moe** [FabricMC/fabric-api#1526] [Mod Dev Support] [answered]
I've got a Mixin right here that overwrites the entire tick method of the BeaconBlockEntity class with a slightly modified version of it. Unfortunately, keeping the "static" keyword makes it say "Public static members are not allowed in Mixin classes". Removing the static keyword makes it crash upon startup. 
Changing "public" to "private" removes the error, but still crashes upon startup.
Is it even possible to overwrite a public static method?

`
@Mixin(BeaconBlockEntity.class)
public class BeaconBlockEntityMixin {
    private static void setNewField_19178(List<BeaconBlockEntity.BeamSegment> field_19178, BeaconBlockEntity blockEntity) {
        ((BeaconBlockEntityAccessor)blockEntity).setField_19178(field_19178);
    }
    private static void setNewMinY(int minY, BeaconBlockEntity blockEntity) {
        ((BeaconBlockEntityAccessor)blockEntity).setMinY(minY);
    }
    private static void setNewLevel(int level, BeaconBlockEntity blockEntity) {
        ((BeaconBlockEntityAccessor)blockEntity).setLevel(level);
    }
    private static void setNewBeamSegments(List<BeaconBlockEntity.BeamSegment> beamSegments, BeaconBlockEntity blockEntity) {
        ((BeaconBlockEntityAccessor)blockEntity).setBeamSegments(beamSegments);
    }

    public static void tick(World world, BlockPos pos, BlockState state, BeaconBlockEntity blockEntity) {
        int minY2 = ((BeaconBlockEntityAccessor)blockEntity).getMinY();
        List<BeaconBlockEntity.BeamSegment> field_191782 = ((BeaconBlockEntityAccessor)blockEntity).getField_19178();
        int level2 = ((BeaconBlockEntityAccessor)blockEntity).getLevel();
        List<BeaconBlockEntity.BeamSegment> beamSegments2 = ((BeaconBlockEntityAccessor)blockEntity).getBeamSegments();
        StatusEffect primary2 = ((BeaconBlockEntityAccessor)blockEntity).getPrimary();
        StatusEffect secondary2 = ((BeaconBlockEntityAccessor)blockEntity).getSecondary();

        int i = pos.getX();
        int j = pos.getY();
        int k = pos.getZ();
        BlockPos blockPos2;
        if (minY2 < j) {
            blockPos2 = pos;
            setNewField_19178(Lists.newArrayList(), blockEntity);
            field_191782 = Lists.newArrayList();
            setNewMinY(pos.getY() - 1, blockEntity);
            minY2 = pos.getY() - 1;
        } else {
            blockPos2 = new BlockPos(i, minY2 + 1, k);
        }

        BeaconBlockEntity.BeamSegment beamSegment = field_191782.isEmpty() ? null : (BeaconBlockEntity.BeamSegment)field_191782.get(field_191782.size() - 1);
        float[] color2 = ((BeamSegmentAccessor)beamSegment).getColor();
        int l = world.getTopY(Heightmap.Type.WORLD_SURFACE, i, k);

        int n;
        for(n = 0; n < 10 && blockPos2.getY() <= l; ++n) {
            BlockState blockState = world.getBlockState(blockPos2);
            Block block = blockState.getBlock();
            if (block instanceof Stainable) {
                float[] fs = ((Stainable)block).getColor().getColorComponents();
                if (field_191782.size() <= 1) {
                    beamSegment = new BeaconBlockEntity.BeamSegment(fs);
                    field_191782.add(beamSegment);
                } else if (beamSegment != null) {
                    if (Arrays.equals(fs, color2)) {
                        ((BeamSegmentInvoker)beamSegment).invokeIncreaseHeight();
                    } else {
                        beamSegment = new BeaconBlockEntity.BeamSegment(new float[]{(color2[0] + fs[0]) / 2.0F, (color2[1] + fs[1]) / 2.0F, (color2[2] + fs[2]) / 2.0F});
                        field_191782.add(beamSegment);
                    }
                }
            } else {
                if (beamSegment == null || blockState.getOpacity(world, blockPos2) >= 15 && !blockState.isIn(DataTags.BEACON_PASS_THROUGH_BLOCKS)) {
                    field_191782.clear();
                    setNewMinY(l, blockEntity);
                    minY2 = l;
                    break;
                }

                ((BeamSegmentInvoker)beamSegment).invokeIncreaseHeight();
            }

            blockPos2 = blockPos2.up();
            setNewMinY(++minY2, blockEntity);
            ++minY2;
        }

        n = level2;
        if (world.getTime() % 80L == 0L) {
            if (!beamSegments2.isEmpty()) {
                setNewLevel(BeaconBlockEntityInvoker.invokeUpdateLevel(world, i, j, k), blockEntity);
                level2 = BeaconBlockEntityInvoker.invokeUpdateLevel(world, i, j, k);
            }

            if (level2 > 0 && !beamSegments2.isEmpty()) {
                BeaconBlockEntityInvoker.invokeApplyPlayerEffects(world, pos, level2, primary2, secondary2);
                BeaconBlockEntity.playSound(world, pos, SoundEvents.BLOCK_BEACON_AMBIENT);
            }
        }

        if (minY2 >= l) {
            setNewMinY(world.getBottomY() - 1, blockEntity);
            minY2 = world.getBottomY() - 1;
            boolean bl = n > 0;
            setNewBeamSegments(field_191782, blockEntity);
            beamSegments2 = field_191782;
            if (!world.isClient) {
                boolean bl2 = level2 > 0;
                if (!bl && bl2) {
                    BeaconBlockEntity.playSound(world, pos, SoundEvents.BLOCK_BEACON_ACTIVATE);
                    Iterator var17 = world.getNonSpectatingEntities(ServerPlayerEntity.class, (new Box((double)i, (double)j, (double)k, (double)i, (double)(j - 4), (double)k)).expand(10.0D, 5.0D, 10.0D)).iterator();

                    while(var17.hasNext()) {
                        ServerPlayerEntity serverPlayerEntity = (ServerPlayerEntity)var17.next();
                        Criteria.CONSTRUCT_BEACON.trigger(serverPlayerEntity, level2);
                    }
                } else if (bl && !bl2) {
                    BeaconBlockEntity.playSound(world, pos, SoundEvents.BLOCK_BEACON_DEACTIVATE);
                }
            }
        }

    }
}`

**Comment by Chocohead** [comment, marked as answer]
You should be able to overwrite public static methods like any other, you'll have to post the crash you're getting from trying to do so. Although in your particular case I can't see what you're changing for whether you need to overwrite the whole method in the first place (as it does prevent any other mod from mixing into the method)

**Reply by crosby-moe** [reply]
In the ide, it underlines the whole method and says "Public static members are not allowed in Mixin classes", and when i run it, it says my mixin class "contains non-private static method tick(Lnet/minecraft/world/World;Lnet/minecraft/util/math/BlockPos;Lnet/minecraft/block/BlockState;Lnet/minecraft/block/entity/BeaconBlockEntity;)V"

**Reply by crosby-moe** [reply]
right so someone somewhere should write that overwriting public static methods requires @Overwrite to work
